### PR TITLE
feat(submission_api): allow students to view output of CE submissions

### PR DESCRIPTION
**Pull request by Bogay**
Backend part of Normal-OJ/Normal-OJ#35

### CODE REVIEW

File: mongo/submission.py [hunk 1]
```diff
@@ -745,12 +751,18 @@ def own_permission(self, user) -> Permission:
...
+                req=Problem.Permission.VIEW,
+        ):
             cap = self.Permission.OTHER
         else:
             cap = self.Permission(0)
 
+        CE = 2
+        if self.status == CE and self.user.username == user.username:
+            cap |= self.Permission.VIEW_OUTPUT
+
```
Comment (ra890927) [inline mongo/submission.py:765]: Can this section put in method `own_permission` ?

**Review by ra890927 — APPROVED** [review]

File: mongo/submission.py [hunk 2]
```diff
@@ -745,12 +751,18 @@ def own_permission(self, user) -> Permission:
...
+                user=user,
+                req=Problem.Permission.VIEW,
+        ):
             cap = self.Permission.OTHER
         else:
             cap = self.Permission(0)
 
+        CE = 2
+        if self.status == CE and self.user.username == user.username:
+            cap |= self.Permission.VIEW_OUTPUT
```
Comment (uier) [inline mongo/submission.py:764]: `self.user.username == user.username` seems redundant, we have defined this as `STUDENT` in above.

Also, does it imply that `MANAGER`/teacher cannot view output?

Comment (Bogay) [reply mongo/submission.py:764]: > `self.user.username == user.username` seems redundant, we have defined this as `STUDENT` in above.

So you suggest to move CE check to the student block?

> 
> Also, does it imply that `MANAGER`/teacher cannot view output?

No, `MANAGER` can view ouput unconditionally, no extra check needed.

Comment (uier) [reply mongo/submission.py:764]: 1. Nah, I suggest to keep CE here and replace `self.user.username == user.username` with `STUDENT`

2. Got it, thanks

**Comment by Bogay** [comment]
根據之前和 @as535364 的討論，可能會多加上允許部分測資公開（不論 CE 與否）的功能，討論完成前先改 Draft。

**Comment by Bogay** [comment]
> 根據之前和 @as535364 的討論，可能會多加上允許部分測資公開（不論 CE 與否）的功能，討論完成前先改 Draft。

為了不要 block 目前繼續寫 submission 測試，討論移到 #207，之後另開 PR

**Review by as535364 — APPROVED** [review]

File: mongo/submission.py [hunk 3]
```diff
@@ -745,12 +751,19 @@ def own_permission(self, user) -> Permission:
...
+        elif Problem(self.problem).permission(
+                user=user,
+                req=Problem.Permission.VIEW,
+        ):
             cap = self.Permission.OTHER
         else:
             cap = self.Permission(0)
 
+        # students can view outputs of their CE submissions
+        CE = 2
```
[the remark this answers is not in the document]
Comment (Bogay) [reply mongo/submission.py:763]: see #264

**Pull request merged**